### DISCUSSION TITLE
Porting mmuart example to custom board; MSS_UART_tx_complete never returns non-zero value

### DISCUSSION THREAD

**Discussion by artruder** [polarfire-soc/polarfire-soc-documentation#186] [Bare metal embedded software] [answered]
Hi,

I am trying to port the existing mmuart example from https://github.com/polarfire-soc/polarfire-soc-bare-metal-examples for our custom module. I managed to adjust everything for our board and the polled transfer completes without problems. Since our board uses only the MMUART 0 I understand that I need to enable the global interrupt in u54_1 in order to get UART output from u54_1 when using interrupt based methods.
I have added the following lines to the example code:
`    __enable_irq();
    PLIC_init(); // add this line

    /* Bring all the MMUARTs out of Reset */
    (void) mss_config_clk_rst(MSS_PERIPH_CFM, (uint8_t) 1, PERIPHERAL_ON);

    /* All clocks ON */

    MSS_UART_set_rx_handler(&g_mss_uart0_lo, uart0_rx_handler,
            MSS_UART_FIFO_SINGLE_BYTE);

//    MSS_UART_enable_local_irq(&g_mss_uart0_lo);
    PLIC_EnableIRQ(MMUART0_PLIC_77); //add this line

    /* Demonstrating polled MMUART transmission */
    MSS_UART_polled_tx(&g_mss_uart0_lo,g_message1,
            sizeof(g_message1));

    /* Demonstrating interrupt method of transmission */
    MSS_UART_irq_tx(&g_mss_uart0_lo, g_message2,
            sizeof(g_message2));

    /* Makes sure that the previous interrupt based transmission is completed
     * Alternatively, you could register TX complete handler using
     * MSS_UART_set_tx_handler() */
    while (0u == MSS_UART_tx_complete(&g_mss_uart0_lo)) {
        ;
    }`

Pausing the program I can see that it is stuck in the while loop at the end of the snippet I posted since this condition: TX_COMPLETE == this_uart->tx_buff_size in the MSS_UART_tx_complete is never met.

Any pointers or suggestions to what I am missing are very much appreciated.

**Comment by AntonKrug** [comment]
Probably there are multiple reasons why it can get stuck waiting on TX_COMPLETE.

But for me it happened on Mi-V RV32 projects when I had wrong memory map. My base addresses of peripherals in the software and where they were mapped in HW were not matching.

**Comment by artruder** [comment]
I thought that would be handled by the mpfs_configuration_generator.py script which is run in the pre-build step. I have created a new entry under the boards folder for our custom board and added the generated MSS config .xml which seems to generate all of the necessary files. I am using the latest Libero SoC 2022.2 and SoftConsole 2021.3. 
Now I have noticed that there is a newer version of the python script available in this repository: https://github.com/polarfire-soc/platform/blob/main/soc_config_generator/mpfs_configuration_generator.py -> 0.6.4 whereas the mmuart example uses 0.6.3. I will try to see if that makes a difference as well.
Furthermore I tried checking the base address of the mmuart0 peripheral, but the link in the TRM leads to a document not found error on the Microchip download page. I found the link to the Polarfire SoC device register map here in the MSS Memory Map chapter: https://www.microsemi.com/document-portal/doc_download/1245725-polarfire-soc-fpga-mss-technical-reference-manual

**Reply by AntonKrug** [reply]
Yes the idea is that the python script takes the XML generated by PolarFire SoC MSS Configurator and produces correct C/H files. Many changes in the PolarFire SoC MSS Configurator can be applied just on a firmware level, however some changes require to re-run the Libero flow as well. Currently I'm not aware of any indication from the tools when this is needed, so if you expirence problems I would do that blindly (and maybe uncesary step) just to be on the safe side.

You mean the link on page 127?
"Peripherals address space, see PolarFire SoC Device Register Map."

I think it can be found on Microchip website now:
https://www.microchip.com/en-us/products/fpgas-and-plds/system-on-chip-fpgas/polarfire-soc-fpgas#Documentation

Datasheet -> PolarFire SoC FPGA Register Map v5

Now only guessing, but another reason maybe could be if the software has right base address, but in the MSS the peripheral is disabled?

**Comment by artruder** [comment]
Thanks for the link regarding the register map, that one works.

The peripheral is enabled in the MSS configurator and the base address should also be fine because otherwise I would not see any output from UART, but I can see the first messages printed with the polled transfer method from e51 and u54 (I have not adjusted the actual messages, so in my case UART1 is UART0)
![image](https://user-images.githubusercontent.com/34745788/186385557-9de0165e-33ef-46b4-a9e1-9df987b3272f.png)

The design is freshly generated with Libero SoC so the XML should be up to date.

**Reply by AntonKrug** [reply]
"Since our board uses only the MMUART 0" could it be that the code you are using is trying to use other MMUARTs and that's why the driver is stuck? Not because of the MMUART0, but because of the other MMUARTs.

**Comment by nitindeshpande** [comment, marked as answer]
@artruder 
In the code snippet you shared, I dont see a call to 
MSS_UART_enable_irq(&g_mss_uart0_lo, (MSS_UART_RBF_IRQ | MSS_UART_TBE_IRQ));

Also the PLIC interrupt priority isnt set
 PLIC_SetPriority(MMUART0_PLIC, EXTERNAL_IRQ_PRIORITY);
 PLIC_SetPriority_Threshold(PLIC_PRIORITY_THRESHOLD);

Could you please try that?

**Reply by artruder** [reply]
@nitindeshpande 
@AntonKrug 
Thanks, that put me in the right direction. I needed to add this line: PLIC_SetPriority(MMUART0_PLIC_77, 1); and now the interrupt based transfers are also working.

Thank you very much for your quick help!
